packages/core/src/anchor/processing-loop.ts

```diff
@@ -0,0 +1,124 @@
+ * It postpones the action to a later stage for whatever reason, but mostly because the expected value is not available yet.
+ *
+ * Unlike Promise that encapsulation fulfillment (resolve or reject) in `new Promise()`,
+ * `Deferred` allows you to return promise, and then fulfill it manually.
+ *
+ * ```typescript
+ * const promiseLike = new Deferred<T>() // Feel free to do whatever you like with it
+ * // When you _want_ the promise to resolve, just pass a value:
+ * promiseLike.resolve(value)
+ * // Same with rejection:
+ * promiseLike.reject(new Error("I am done"))
+ * ```
+ */
+export class Deferred<T = void> implements PromiseLike<T> {
+  /**
+   * Resolve the promise manually.
+   */
+  readonly resolve: (t: T) => void
+
+  /**
+   * Reject the promise.
+   */
+  readonly reject: (e: Error) => void
+
+  // Parts of Promise API
+  readonly then: Promise<T>['then']
+  readonly catch: Promise<T>['catch']
+
+  constructor() {
+    let dResolve: Deferred<T>['resolve']
+    let dReject: Deferred<T>['reject']
+
+    const promise = new Promise<T>((resolve, reject) => {
+      dResolve = resolve
+      dReject = reject
+    })
+    this.resolve = dResolve
+    this.reject = dReject
+    this.then = promise.then.bind(promise)
+    this.catch = promise.catch.bind(promise)
+  }
+}
+
+export class ProcessingLoop<T> {
+  /**
+   * Source of entries we process in a loop.
+   */
+  private readonly source: AsyncGenerator<T>
+  /**
+   * External action that processes an entry.
+   */
+  private readonly handleValue: (value: T) => Promise<void>
+  /**
+   * `for await ()` loop wrapped in a function. It resolves, when `source` reports it has no entries.
+   * If absent, i.e. `undefined`, it means, we have not started processing the entries.
+   */
+  #processing: Promise<void> | undefined
+  /**
+   * Used to intercept fulfillment inside `#processing`. Node.js forces you to handle Promise errors by
+   * issuing a warning to console. Passing an error to an instance of `Deferred` solves that.
+   */
+  #whenComplete: Deferred
+
+  /**
+   * Stop promises in `#processing` by issuing abort signal.
+   */
+  #abortController: AbortController
+
+  constructor(source: ProcessingLoop<T>['source'], onValue: ProcessingLoop<T>['handleValue']) {
+    this.source = source
+    this.handleValue = onValue
+    this.#processing = undefined
+    this.#whenComplete = new Deferred()
+    this.#abortController = new AbortController()
+  }
+
+  start() {
+    const rejectOnAbortSignal = new Promise<IteratorResult<T>>((resolve) => {
+      if (this.#abortController.signal.aborted) {
+        resolve({ done: true, value: undefined })
+      }
+      const done = () => {
+        this.#abortController.signal.removeEventListener('abort', done)
+        resolve({ done: true, value: undefined })
+      }
+      this.#abortController.signal.addEventListener('abort', done)
+    })
+    const processing = async (): Promise<void> => {
```

Comment:
I'm trying to understand the degree of concurrency that happens with the new polling system.  Let me see if I'm understanding correctly.  I see that when we query entries from the AnchorRequestStore, we use a batch size of 10.  So we fetch 10 entries to process from the store at once, but if I'm reading this properly, we still process them one at a time since this processing loop waits for each value to be handled before starting processing on the next value, is that right?

I think we're likely going to need more concurrency here to keep up with the incoming write load, especially with our planned performance targets for this season, but I'm fine with leaving that for follow-up work, just trying to make sure I understand the current state of things

Reply:
Yes, it is one streamid at a time. You are correct. We should explore how to make it more concurrent definitely. We have talked about it before. If I remember correctly, the decision was to postpone more advanced polling to the next iteration, after we confirm the off-chain memory mechanism works as a whole.